src/chunk.c

```diff
@@ -4613,3 +4618,70 @@ ts_chunk_attach_osm_table_chunk(PG_FUNCTION_ARGS)
 
 	PG_RETURN_BOOL(ret);
 }
+
+static ScanTupleResult
+chunk_tuple_osm_chunk_found(TupleInfo *ti, void *arg)
+{
+	bool isnull;
+
+	Datum osm_status_datum = slot_getattr(ti->slot, Anum_chunk_status, &isnull);
+	Assert(!isnull);
+	int32 osm_status = DatumGetInt32(osm_status_datum);
+	bool is_dropped = DatumGetBool(slot_getattr(ti->slot, Anum_chunk_dropped, &isnull));
+	if (is_dropped || !ts_flags_are_set_32(osm_status, CHUNK_STATUS_FOREIGN))
+		return SCAN_CONTINUE;
+
+	int *chunk_id = (int *) arg;
+	Datum chunk_id_datum = slot_getattr(ti->slot, Anum_chunk_id, &isnull);
+	Assert(!isnull);
+	*chunk_id = DatumGetInt32(chunk_id_datum);
+
+	return SCAN_DONE;
+}
+
+/* get OSM chunk id associated with the hypertable */
+int
+ts_chunk_get_osm_chunk_id(int hypertable_id)
+{
+	int chunk_id = INVALID_CHUNK_ID;
+	ScanKeyData scankey[2];
+	int32 osm_chunk_status = CHUNK_STATUS_FOREIGN;
+	Catalog *catalog = ts_catalog_get();
+	int num_found;
+	ScannerCtx scanctx = {
+		.table = catalog_get_table_id(catalog, CHUNK),
+		.index = catalog_get_index(catalog, CHUNK, CHUNK_STATUS_HYPERTABLE_ID_INDEX),
+		.nkeys = 2,
+		.scankey = scankey,
+		.data = &chunk_id,
+		.filter = NULL,
+		.tuple_found = chunk_tuple_osm_chunk_found,
+		.lockmode = AccessShareLock,
+		.scandirection = ForwardScanDirection,
+	};
+
+	/*
+	 * Perform an index scan on status + hypertable ID
+	 */
+	ScanKeyInit(&scankey[0],
+				Anum_chunk_status_hypertable_id_idx_status,
+				BTEqualStrategyNumber,
+				F_INT4EQ,
+				Int32GetDatum(osm_chunk_status));
+	ScanKeyInit(&scankey[1],
+				Anum_chunk_status_hypertable_id_idx_hypertable_id,
+				BTEqualStrategyNumber,
+				F_INT4EQ,
+				Int32GetDatum(hypertable_id));
+
+	num_found = ts_scanner_scan(&scanctx);
```

Comment:
Do you want to simplify this to an index-only scan? You'll need to make a covering index  with `include (id)`, set `.table = NULL` and `.want_itup` in the context, use `ts_scanner_next` to get the index tuple, and then `index_getattr()` to get the chunk id from the index tuple directly.

Reply:
This would be less I/O and less code.